src/AutoRest.CSharp/Common/AutoRest/Plugins/GeneratedCodeWorkspace.cs

```diff
@@ -170,5 +171,236 @@ public async Task<CSharpCompilation> GetCompilationAsync()
+            }
+
+            public override SyntaxNode? VisitEnumDeclaration(EnumDeclarationSyntax node)
+            {
+                node = (EnumDeclarationSyntax)base.VisitEnumDeclaration(node)!;
+                if (IsPublic(node.Modifiers) && !_modelsToKeep.Contains(node.Identifier.ToString()))
+                    _declarations.Add(node);
+                return node;
+            }
+        }
+
+        private class PublicMemberVisitor : CSharpSyntaxRewriter
+        {
+            private List<SyntaxNode> _members = new();
+
+            public IEnumerable<SyntaxNode> PublicMembers => _members;
+
+            public override SyntaxNode? VisitClassDeclaration(ClassDeclarationSyntax node)
+            {
+                node = (ClassDeclarationSyntax)base.VisitClassDeclaration(node)!;
+                if (IsPublic(node.Modifiers))
+                {
+                    _members.Add(node); // add myself
+                }
+                return node;
+            }
+
+            public override SyntaxNode? VisitPropertyDeclaration(PropertyDeclarationSyntax node)
+            {
+                node = (PropertyDeclarationSyntax)base.VisitPropertyDeclaration(node)!;
+                if (IsPublic(node.Modifiers))
+                {
+                    _members.Add(node);
+                }
+                return node;
+            }
+
+            public override SyntaxNode? VisitMethodDeclaration(MethodDeclarationSyntax node)
+            {
+                node = (MethodDeclarationSyntax)base.VisitMethodDeclaration(node)!;
+                if (IsPublic(node.Modifiers))
+                    _members.Add(node);
+                return node;
+            }
+
+            public override SyntaxNode? VisitConstructorDeclaration(ConstructorDeclarationSyntax node)
+            {
+                node = (ConstructorDeclarationSyntax)base.VisitConstructorDeclaration(node)!;
+                if (IsPublic(node.Modifiers))
+                    _members.Add(node);
+                return node;
+            }
+        }
+
+        private static bool IsPublic(SyntaxTokenList tokenList)
+            => tokenList.Any(token => token.IsKind(SyntaxKind.PublicKeyword));
+
+        private static bool IsStatic(SyntaxTokenList tokenList)
+            => tokenList.Any(token => token.IsKind(SyntaxKind.StaticKeyword));
+
+        private async Task<IEnumerable<SyntaxNode>> GetAllDeclaredModels(ImmutableHashSet<string> modelsToKeep)
+        {
+            var classVisitor = new PublicDefinitionVisitor(modelsToKeep);
+
+            foreach (var document in _project.Documents)
+            {
+                if (IsGeneratedDocument(document))
+                {
+                    var root = await document.GetSyntaxRootAsync();
+                    classVisitor.Visit(root);
+                }
+            }
+
+            return classVisitor.ModelDeclarations;
+        }
+
+        private async IAsyncEnumerable<SyntaxNode> TraverseAllPublicModels()
```

Comment:
This method misses lots of cases, for example:
- If model 'A' is publicly referenced by model 'B' and model 'B' is publicly referenced by rest operation, both should be kept public
- If collection of models is publicly referenced by rest operation, it should be kept public
- If generic parameter constraint publicly references model, it should be kept public

and so on, and any combination of this, e.g.: `GetSomething<T>() where T: IEnumerable<SomeModel>`

It also works with syntax nodes rather than symbols, which is incorrect. I'll look a bit more into this today.

Reply:
I believe the first two cases are covered by the current implementation, since this is also adding types in the type arguments recursively.
I did not find an example of the third case in mgmt generated code - do we have this case? I have generated all the existing RPs on this PR: https://github.com/Azure/azure-sdk-for-net/pull/27204
And a case of discriminator is also missing in the current implementation, like we have a model A returned by an operation, and it has multiple derived types from the discriminator and those derived types should be marked as public as well.